File: solutions/typescript/minesweeper/1/minesweeper.ts
export function annotate(field: string[]): string[] {
  const nrow = field.length
  const ncol = field[0]?.length ?? 0
  for (const row of field) {
    if (row.length !== ncol) {
      throw new Error("Uneven fields")
    }
  }
  const count = (i: number, j: number): number => {
    let mines = 0
    for (let x = Math.max(0, i - 1); x <= Math.min(i + 1, nrow - 1); ++x) {
      for (let y = Math.max(0, j - 1); y <= Math.min(j + 1, ncol - 1); ++y) {
        if (field[x][y] === '*') ++mines
      }
    }
    return mines
  }
  const buf = Array.from({length: ncol})
  for (let i = 0; i < nrow; ++i) {
    for (let j = 0; j < ncol; ++j) {
      if (field[i][j] === '*') {
        buf[j] = '*'
        continue
      }
      const mines = count(i, j)
      buf[j] = mines === 0 ? ' ' : String(mines)
    }
    field[i] = buf.join('')
  }
  return field
}

File: solutions/kotlin/minesweeper/1/src/main/kotlin/Minesweeper.kt
import kotlin.math.max
import kotlin.math.min

data class MinesweeperBoard(val board: List<String>) {
    val annotated: List<String> = board.let { board ->
        val nrows = board.size
        val ncols = if (nrows == 0) { 0 } else { board[0].length }
        require(board.all { row -> row.length == ncols })
        val isMine: (Int, Int) -> Boolean = { rowIdx, colIdx -> board[rowIdx][colIdx] == '*' }
        val adjacentMines: (Int, Int) -> Int = { rowIdx, colIdx ->
            (max(0, rowIdx - 1)..min(rowIdx + 1, nrows - 1)).flatMap { adjRowIdx ->
                (max(0, colIdx - 1)..min(colIdx + 1, ncols - 1)).map { adjColIdx -> 
                    adjRowIdx to adjColIdx
                }
            }.filter { (adjRowIdx, adjColIdx) -> isMine(adjRowIdx, adjColIdx) }.count()
        }
        board.mapIndexed { rowIdx, row ->
            row.asSequence().mapIndexed { colIdx, cell ->
                if (cell == '*') { '*' } else {
                    adjacentMines(rowIdx, colIdx).let { adjMineCount -> 
                        if (adjMineCount == 0) { ' ' } else { adjMineCount.digitToChar() }
                    }
                }
            }.joinToString(separator = "")
        }
    }

    fun withNumbers(): List<String> = annotated
}